client/src/components/ExpenseForm.tsx

```diff
@@ -0,0 +1,272 @@
+};
+
+const ExpenseForm: React.FC = () => {
+  const { variant } = useAppSelector(state => state.modal);
+  const { currentExpenseId } = useAppSelector(state => state.expense);
+
+  const [ currentExpense, setCurrentExpense ] = useState<IExpense | null>(null);
+
+  const [formUser, setFormUser] = useState('');
+  const [formTitle, setFormTitle] = useState('');
+  const [formAmount, setFormAmount] = useState(0);
+  const [formCategory, setFormCategory] = useState('');
+  const [formDate, setFormDate] = useState(
+    normalizeDateForDb(new Date().toJSON())
+  );
+  const [formNote, setFormNote] = useState('');
+  const [isLoading, setIsLoading] = useState(false);
+  const [isProcessing, setIsProcessing] = useState(false);
+
+  const dispatch = useAppDispatch();
+
+  const handleFormSubmit = (event: FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+    setIsProcessing(true);
+
+    if (variant === 'new') {
+      addExpense({
+        user: formUser,
+        title: formTitle,
+        category: formCategory,
+        amount: formAmount,
+        date: formDate,
+        note: formNote,
+      })
+        .then(() => {
+          dispatch(setModalState({ variant: null }));
+          dispatch(fetchExpenses());
+        })
+        .catch(() => {
+          setIsProcessing(false);
+        });
+    }
+
+    if (variant === 'edit') {
+      if (!currentExpense) {
+        setIsProcessing(false);
+
+        return;
+      }
+
+      patchExpense({
+        id: currentExpense.id,
+        user: formUser,
+        title: formTitle,
+        category: formCategory,
+        amount: formAmount,
+        date: formDate,
+        note: formNote,
+      })
+        .then(() => {
+          dispatch(setModalState({ variant: null }));
+          dispatch(fetchExpenses());
+        })
+        .catch(() => {
+          setIsProcessing(false);
+        });
+    }
+  }
+
+  const handleCancel = () => {
+    dispatch(setModalState({ variant: null }));
+  };
+
+  useEffect(() => {
+    if (!currentExpenseId) {
+      return;
+    }
+
+    setIsLoading(true);
+
+    getExpenseById(currentExpenseId)
+      .then(result => setCurrentExpense(result))
+      .finally(() => setIsLoading(false));
+  }, [currentExpenseId]);
+
+  useEffect(() => {
+    if (!currentExpense) {
+      return;
+    }
+
+    const {
+      user,
+      title,
+      category,
+      amount,
+      date,
+      note,
+    } = currentExpense;
+
+    setFormUser(user);
+    setFormTitle(title);
+    setFormCategory(category);
+    setFormAmount(amount);
+    setFormDate(date);
+    setFormNote(note);
+  }, [currentExpense]);
+
+  return (
+    <>
+      {isLoading && <Loader className="mx-auto is-size-1" />}
+      {!isLoading && (
+        <form onSubmit={handleFormSubmit}>
+        <Form.Field>
+          <Form.Label>User</Form.Label>
+          <Form.Control>
+            <Form.Input
+              type="text"
+              value={formUser}
+              onChange={(e) => {
```

Comment:
Add changes all over the code.
<img width="1092" alt="Screenshot 2022-09-26 at 13 14 03" src="https://user-images.githubusercontent.com/17098736/192251854-bdf8c92a-2c3e-4001-b30c-d7915577947e.png">